### PR TITLE
CUDAGraph support for SimpleFSDP and TP

### PR DESCRIPTION
**Pull request by BoyuanFeng**
## Features
- [x] Support SimpleFSDP and TP
- [x] Support static input indices to reduce copy
- [x] Support memory reuse to reduce memory consumption
- [x] Cleanup cudagraph when training finishes to avoid nccl hang from destroy_process_group

Command:
```
NCCL_GRAPH_REGISTER=0 NGPU=8 TRAIN_FILE=torchtitan.experiments.compiler_toolkit.train CONFIG_FILE=./torchtitan/models/llama3/train_configs/debug_model.toml ./run_train.sh --model.name compiler_toolkit.llama3 --parallelism.data_parallel_shard_degree=2 --parallelism.tensor_parallel_degree=4  --job.custom_config_module=torchtitan.experiments.compiler_toolkit.job_config --compile.passes cudagraph
```

Note: we use `NCCL_GRAPH_REGISTER=0` due to a known issue that nccl + cudagraphs + expandable segments result in IMA. https://github.com/pytorch/pytorch/issues/158029

[trace](https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces%2Ftree%2Fshared_trace%2Fboyuan_e1ef464b-ee61-4c61-82e5-f7a485e561bf_rank0_trace.json)

## Result

**Numerics:**
Achieved bitwise equivalence w/ and w/o cudagraph pass on llama3.1-8B AND llama3.1-70B.

**Performance:**
image

Raw log: [llama3-8b](https://www.internalfb.com/phabricator/paste/view/P2045444190), [llama3-70b](https://www.internalfb.com/phabricator/paste/view/P2045567416)

**Memory:**
On llama3.1-70b, cudagraph takes 6% more memory consumption (143 GiB vs 153 GiB).

A few tricks to reduce memory consumption (use llama3.1-70b w/ cudagraph as an example):
- Start: 161 GiB
- \+ use the same stream for warmup and graph capture of both fwd and bwd: 160 GiB
- \+ warmup in cudagraph memory pool instead of eager memory pool: 153 GiB

**static input copy:**
On llama3.1-70B, for forward, we copy 1 tensor of 128 bytes; for backward, we copy 1 tensor of 0.98 GB. This shows static input indices is handled correctly.

## Followup PR
In the followup PR, I will enable fx graph partition for deepseek v3 https://github.com/pytorch/pytorch/pull/165945.

### CODE REVIEW

File: torchtitan/experiments/compiler_toolkit/cudagraph.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        ]
+        self.cudagraph: Optional[torch.cuda.CUDAGraph] = None
+        self.has_warmup = False
+
+        self.args = None
+        self.output = None
+
+    def copy_static_inputs(self, *args):
+        for i in self.input_indices_to_copy:
+            self.args[i].copy_(args[i])
```
Comment (BoyuanFeng) [inline torchtitan/experiments/compiler_toolkit/cudagraph.py:89]: we could replace this for loop with foreach copy. However, I empirically observed there is only 1 tensor to copy for fwd and 1 tensor to copy for bwd. So no need to add code complexity here.

File: torchtitan/experiments/compiler_toolkit/cudagraph.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            ]
+            self.input_addresses = input_addresses
+
+            self.cudagraph = torch.cuda.CUDAGraph()
+
+            with torch.cuda.graph(
+                self.cudagraph, pool=self.graph_pool, stream=self.stream
+            ):
+                # `output` is managed by pytorch's cudagraph pool
+                self.output = self.runnable(*args)
```
Comment (BoyuanFeng) [inline torchtitan/experiments/compiler_toolkit/cudagraph.py:132]: we could potentially use weakref for output tensor to reduce memory. Will do in a followup pr.

File: torchtitan/experiments/compiler_toolkit/cudagraph.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            # across eager memory pool and cudagraph memory pool.
+            with _use_cuda_memory_pool_manager(device, self.graph_pool, self.stream):
+                out = self.runnable(*args)
+            return out
+
+        if self.cudagraph is None:
+            self.args = args
+            input_addresses = [
+                x.data_ptr() if isinstance(x, torch.Tensor) else None for x in args
+            ]
```
Comment (eellison) [inline torchtitan/experiments/compiler_toolkit/cudagraph.py]: I guess we're assuming that the non tensor inputs are the same every time ? Should we just assert they're all tensors if we're not handling the other cases ?

File: torchtitan/experiments/compiler_toolkit/cudagraph.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+            with torch.cuda.graph(
+                self.cudagraph, pool=self.graph_pool, stream=self.stream
+            ):
+                # `output` is managed by pytorch's cudagraph pool
+                self.output = self.runnable(*args)
+
+        self.copy_static_inputs(*args)
+        self.cudagraph.replay()
+        return self.output
```
Comment (eellison) [inline torchtitan/experiments/compiler_toolkit/cudagraph.py:139]: The persistent input and output is not good for memory, as you've commented.

File: torchtitan/experiments/compiler_toolkit/cudagraph.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            for i, inp in enumerate(example_inputs)
+            if isinstance(inp, torch.Tensor) and i not in self.static_input_indices
+        ]
+        self.cudagraph: Optional[torch.cuda.CUDAGraph] = None
+        self.has_warmup = False
+
+        self.args = None
+        self.output = None
+
+    def copy_static_inputs(self, *args):
```
Comment (eellison) [inline torchtitan/experiments/compiler_toolkit/cudagraph.py]: if any of the static inputs changes, you'll get silent incorrectness. you might consider at least a config to check this.

File: torchtitan/experiments/compiler_toolkit/cudagraph.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        pass
+
+    return dummy_graph, graph_pool, graph_capture_stream
+
+
+(
+    _global_dummy_graph,
+    _global_graph_pool,
+    _global_graph_capture_stream,
+) = init_global_graph_pool()
```
Comment (eellison) [inline torchtitan/experiments/compiler_toolkit/cudagraph.py:56]: does this work when backward is on a separate stream ? or not an issue?

File: run_train.sh [hunk 7]
```diff
@@ -19,13 +19,18 @@ DRY_RUN=${DRY_RUN:-0}
 
 TORCHFT_LIGHTHOUSE=${TORCHFT_LIGHTHOUSE:-"http://localhost:29510"}
 
+# need to turn off expandable segments when using cudagraph, since
+# it does not work with cg and nccl yet.
+# https://github.com/pytorch/pytorch/issues/158029
```
Comment (eellison) [inline run_train.sh]: Can we turn this off only when using cudagraph ?

File: torchtitan/experiments/compiler_toolkit/cudagraph.py (hunk 6 above)
Comment (BoyuanFeng) [reply torchtitan/experiments/compiler_toolkit/cudagraph.py:56]: IIUC, this is not an issue currently. since fwd and bwd are on the same cuda stream by default.

cudagraph trees has used the same graph capture stream for both fwd and bwd. 
https://github.com/pytorch/pytorch/blob/7a928397cda89b71c24b0efe9db6df7fb04a46cb/torch/_inductor/cudagraph_trees.py#L1945

File: run_train.sh (hunk 7 above)
Comment (BoyuanFeng) [reply run_train.sh]: currently it's on by default. when using cudagraph, we need to explicitly turn it off with `USE_EXPANDABLE_SEGMENTS=False [other commands]`.

File: torchtitan/experiments/compiler_toolkit/cudagraph.py (hunk 3 above)
Comment (BoyuanFeng) [reply torchtitan/experiments/compiler_toolkit/cudagraph.py]: IIUC, there would only be tensor and symint (for moe layer). let me add assertion

File: torchtitan/experiments/compiler_toolkit/cudagraph.py (hunk 4 above)
Comment (BoyuanFeng) [reply torchtitan/experiments/compiler_toolkit/cudagraph.py:139]: yes will add in the next pr.

File: torchtitan/experiments/compiler_toolkit/cudagraph.py (hunk 5 above)
Comment (BoyuanFeng) [reply torchtitan/experiments/compiler_toolkit/cudagraph.py]: yes added

File: torchtitan/experiments/compiler_toolkit/cudagraph.py (hunk 3 above)
Comment (BoyuanFeng) [reply torchtitan/experiments/compiler_toolkit/cudagraph.py]: there are also `rng_state: torch._C.Generator`, used by 

`graphsafe_run_with_rng_state_2 = torch.ops.higher_order.graphsafe_run_with_rng_state(torch.ops.aten._scaled_dot_product_flash_attention.default, transpose_20, transpose_21, transpose_22, 0.0, True, scale = 0.25, rng_state = fwd_rng_state_2);`

See the last 3 args in [P2047035404](https://www.internalfb.com/phabricator/paste/view/P2047035404)

File: torchtitan/experiments/compiler_toolkit/graph_utils.py [hunk 8]
```diff
@@ -323,6 +360,10 @@ def get_compiler_passes_from_config(model: torch.nn.Module, job_config: JobConfi
         else:
             compiler_passes.append(AVAILABLE_COMPILER_PASSES[pass_name])
 
+    if use_cudagraph:
```
Comment (yiming0416) [inline torchtitan/experiments/compiler_toolkit/graph_utils.py]: you have checked cudagraph to be the lass pass in `validate_pass_names()`, so no need for `use_cudagraph` here

File: torchtitan/train.py [hunk 9]
```diff
@@ -703,6 +704,15 @@ def close(self) -> None:
         if hasattr(self, "metrics_processor") and self.metrics_processor:
             self.metrics_processor.close()
 
+        # Note [explicit cudagraph close]
+        # cudagraph holds reference to nccl which prevents destroy nccl
+        # group. so we need to explicitly delete cudagraph which is held
+        # in joint_graph_module. An explicit gc.collect() is necessary
+        # to clean up reference cycles.
+        for part in self.model_parts:
+            part.joint_graph_module = None
```
Comment (yiming0416) [inline torchtitan/train.py]: I think `joint_graph_module` only exists for compiler toolkit experiments. So for other experiments or training runs `part` won't have `joint_graph_module`

Comment (BoyuanFeng) [reply torchtitan/train.py]: we can only cleanup if it has the joint_graph_module

**Review by yiming0416 — APPROVED** [review]
experiments part LGTM

File: torchtitan/train.py [hunk 10]
```diff
@@ -703,6 +704,16 @@ def close(self) -> None:
         if hasattr(self, "metrics_processor") and self.metrics_processor:
             self.metrics_processor.close()
 
+        # Note [explicit cudagraph close]
+        # cudagraph holds reference to nccl which prevents destroy nccl
+        # group. so we need to explicitly delete cudagraph which is held
+        # in joint_graph_module. An explicit gc.collect() is necessary
+        # to clean up reference cycles.
+        for part in self.model_parts:
+            if hasattr(part, "joint_graph_module"):
```
Comment (tianyu-l) [inline torchtitan/train.py]: `joint_graph_module` is exclusively used for `compiler_toolkit` right? If it can't be made general, let's create a `Trainer` subclass to overwrite this method. E.g. https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/torchcomms/train.py

**Review by tianyu-l — COMMENTED** [review]
Sounds great! Had a suggestion.

File: run_train.sh (hunk 7 above)
Comment (ngimel) [reply run_train.sh]: Instead of turning off expandable segments you can turn off nccl memory registration, as the issue suggests

File: torchtitan/train.py (hunk 10 above)
Comment (yiming0416) [reply torchtitan/train.py]: @tianyu-l `Trainer` subclass added in https://github.com/pytorch/torchtitan/pull/2064
